File: 091.LeetCode@133_Clone_Graph/091.LeetCode@133_Clone_Graph.py
from typing import Optional


class Solution:
    def __init__(self):
        # Dictionary to store the cloned nodes
        self.cloned_nodes = {}

    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        # Base case: if the input node is None, return None
        if node is None:
            return None

        # If the node has already been cloned, return the clone
        if node in self.cloned_nodes:
            return self.cloned_nodes[node]

        # Clone the current node (without neighbors initially)
        clone = Node(node.val)
        # Store the clone in the dictionary to avoid cloning the same node again
        self.cloned_nodes[node] = clone

        # Recursively clone all the neighbors and add them to the cloned node's neighbors list
        for neighbor in node.neighbors:
            clone.neighbors.append(self.cloneGraph(neighbor))

        return clone

File: 091.LeetCode@133_Clone_Graph/091.LeetCode@133_Clone_Graph.java
class Solution {
    // HashMap to store the cloned nodes
    private Map<Node, Node> clonedNodes = new HashMap<>();

    public Node cloneGraph(Node node) {
        if (node == null)
            return null;

        // If the node has already been cloned, return the cloned node
        if (clonedNodes.containsKey(node)) {
            return clonedNodes.get(node);
        }

        // Clone the node (without neighbors initially)
        Node clonedNode = new Node(node.val);
        clonedNodes.put(node, clonedNode);

        // Clone the neighbors recursively
        for (Node neighbor : node.neighbors) {
            clonedNode.neighbors.add(cloneGraph(neighbor));
        }

        return clonedNode;
    }
}